### ISSUE TITLE
GFM version 7, suggest things here

### ISSUE THREAD

**Issue by p0nce** [d-gamedev-team/gfm#169]
~~STALLED BECAUSE WAITING ON DERELICTUTIL v3~~
STALLED BECAUSE NO TIME

TODO for v7
- [x] update all wrappers to derelict packages using current DerelictUtil 
- [x] wideint will go in its own independent DUB package (meh, won't do it)
- [x] absByElem for `Vector`
- [x] @fastmath (no need anymore)
- [x] change `length` to `magnitude` and `squaredLength` to `squaredMagnitude`. Better names in general, more fit for completion like in GLM or XNA names (concept then options). Keep them deprecated for now. `length` should be for "container size" in D.
- [x] Include #192 in this major tag (EDIT: no)

**Comment by andrewbenton** [comment]
It would be nice to avoid issues with non-stack allocated sdl objects

**Comment by p0nce** [comment]
What are those issues?

**Comment by andrewbenton** [comment]
The issue that I'm referring to is when you try to use SDL2 / SDL2Window / etc without being wrapped in `scope!T` unless you very carefully manage the scope for those objects.  You end up seeing GC errors and getting a non-zero exit code.  Maybe there's a nice way to do ref-counting for cleanup after all SDL objects are done?

**Comment by p0nce** [comment]
The GFM example use `scoped!T` but there are other ways to release ressource. 
I recommend you simple use `.destroy()` instead, so that such ressource can be fields. The destructors are the ideal place to do that for fields.0

The reason you are seeing GC errors is that if GFM didn't made relying on the GC for cleanup an error, it would be worse like it was before the GC-proof resource class. Things would crash randomly (or not) at exit.

**Comment by claudemr** [comment]
Is it worth adding an abs() function to gfm.math.vector?
I came across that need a couple of times while playing with Vector's and dual-contouring.

**Comment by claudemr** [comment]
I meant absolute value per element. Something like this:
```
static if (isSigned!T)
{
    @property @nogc ref Vector abs() pure nothrow const
    {
        Vector!(T, N) v;
        import std.math : abs;
        mixin(generateLoopCode!("v[@] = abs(this.v[@]);", N)());
        return v;
    }
}
```

**Comment by p0nce** [comment]
Let's call it `absByElem` then?

**Comment by claudemr** [comment]
Have you had issues by using same names as in Phobos?
Using explicit naming (with module name) and aliases can be used to avoid name collision.

Anyway, I don't mind using absByElem, it's fine by me.

As for "length" name, when I did my own vector implementation, I used "magnitude" as the function definition name, and "length" as an alias. Same for "scalarProduct" aliased to "dotProduct" and "vectorProduct" to "crossProduct". I did not want to have to chose between the proper mathematical naming, and the naming developers are used to.

**Comment by p0nce** [comment]
I know there are always ways to avoid name clashes.
At one point gfm:math had "min", "max" and "clamp", it made lots of problems that tend to creep up when improting whole `package.d`.

My opinion here would be to have only one name and one that doesn't offer room for misunderstanding.
If I see "vec.abs" how to know if it's the absolutes value of elements, sum of absolutes value of elements or a vector result?

Your "magnitude" seem better than "length" btw and I will probably change to that name at one point.

**Comment by claudemr** [comment]
Ok, I understand now. It's a good point.

**Comment by claudemr** [comment]
By the way, I have a question about something totally different. I noticed most Vector and Matrix methods (and especially operator ones) have their parameter passed by value. I know using reference parameter is problematic (like "const ref Vector vec") when using constructor argument (hence your D-idiom about "auto ref and not using it").
But, passing by value, doesn't it criple code performance? Especially with 4x4 matrices?
I may have missed a train, maybe the compiler does some kind of optimization, with function inlining or something.

**Comment by p0nce** [comment]
Good question.
I learned about "auto ref" when https://github.com/Dgame submitted the idiom, I'm still not sure where it should be used actually. Can you help?

I've never measured seriously, one time I tried `ref` for passing Vector and it wasn't better. I guess it should be useful
BTW the d-idiom warns about "template bloat" but in practice I don't think this will happen. I should consider auto-ref but as I don't use Matrix a lot in my products it won't be a realistic test.

**Comment by claudemr** [comment]
I dug into dlang forum to see if there were any discussions about this, and I figured it was kind of a polemical question (I've got to find the specific thread about this, it seems Andrei had some specific thoughts about that). Anyway, I'll be happy to help.

I'll make some tests, to see what's going on. I also have to learn more about that "template bloat" problem, in which exact conditions we have to care about it.

**Comment by p0nce** [comment]
> I'll make some tests, to see what's going on. I also have to learn more about that "template bloat" problem, in which exact conditions we have to care about it.

In my opinion, never.
It's a semi-mythical beast that appears with too much different instantiation supposed to make you program compiles slower (this part is true) and execute slower (hasn't seen this in practice, rather the contrary).

**Comment by peerhenry** [comment]
## texture functions
Please add methods to `GLTexture` for the following gl functions:
- `glTexParameterfv(_target, GL_TEXTURE_BORDER_COLOR, color);`
- `glTexParameteri(_target, GL_TEXTURE_COMPARE_MODE, compareMode);`
- `glTexParameteri(_target, GL_TEXTURE_COMPARE_FUNC, compareFunc);`

## diagonal matrices
It would be nice to have a static `diag` method in `Matrix` that takes an n-vector and returns a n x n matrix with the vector on its diagonal. This is useful for example with non-uniform scaling.

**Comment by p0nce** [comment]
All available as v7.0.0

However because of DerelictGL3 changes it is currently impossible to setup which opengl version is loaded
